hyperspy/model.py

```diff
@@ -882,6 +887,249 @@ def _model_function(self, param):
+            "`parameters.free = False`. These " \
+            "parameters are nonlinear:"
+        nonlinear_parameters = self.nonlinear_parameters
+        if nonlinear_parameters:
+            raise AttributeError(not_linear_error + "\n\t" + str(
+        "\n\t".join(str(para) for para in nonlinear_parameters))
+        )
+        if not self.linear_parameters:
+            raise AttributeError("There are no linear components in this model")
+
+
+        self._set_linear_parameters_to_one()
+        self._set_p0()
+        n_free_para = len(self.free_parameters)
+        assert n_free_para > 0, 'Model does not contain any free components!'
+        channels_signal_shape = np.count_nonzero(self.channel_switches)
+        arr_nav_shape = self.axes_manager._navigation_shape_in_array
+        
+        # nav_shape and comp_nav_shape may change with future linear fitting speedups
+        nav_shape = ()
+        comp_nav_shape = ()
+
+        self._component_data = np.zeros(
+            comp_nav_shape + (n_free_para, channels_signal_shape)
+        )
+        self._component_data_fixed = np.zeros(
+            comp_nav_shape + (channels_signal_shape, ))
+
+        self.coefficient_array = np.zeros(nav_shape + (n_free_para,))
+        self._set_twinned_lists()
+        self._assign_twin_value_maps()
+        for component in self:
+            if component.active:
+                self._append_component(component)
+
+        # The linear components
+        self._remove_any_zero_components()
+
+        target_signal = self.signal()[np.where(self.channel_switches)]
+
+        if self.signal.metadata.Signal.binned is True:
+            target_signal = target_signal / self.signal.axes_manager[-1].scale
+
+        target_signal = target_signal - self._component_data_fixed  # - model_constant_term
+
+        component_data_shape = self._component_data.shape
+        sig1Dshape = np.count_nonzero(self.channel_switches)
+
+        # Reshape what may potentially be Signal2D data into a long Signal1D shape
+        target_signal = target_signal.reshape(nav_shape + (sig1Dshape,))
+    
+        #self.coefficient_array[:] = linear_regression(target_signal, self._component_data)
+        if not import_sklearn.sklearn_installed or algorithm=='matrix_inversion':
+            if algorithm == 'ridge_regression':
+                warnings.warn(
+                "Linear fitting is preferably done using the scikit-learn ridge regression code. "
+                "Install scikit-learn (sklearn) to use it. Proceding using a more fragile "
+                "matrix inversion approach."
+                )
+            self.coefficient_array[:] = linear_regression(target_signal, self._component_data)
+        elif algorithm == 'ridge_regression':
+            ridge_regression = import_sklearn.sklearn.linear_model._ridge.ridge_regression
+            self.coefficient_array[:] = ridge_regression(self._component_data.T, target_signal, alpha=0.0, solver="cholesky")
```

Comment:
What about allowing kwargs to be passed to this function, with sensible defaults (e.g. these) if not present in the kwargs dict?

This is a nice PR!

Reply:
Absolutely!

=)